### PR TITLE
Shifts user auth to outside the graphql resolvers

### PR DESCRIPTION
**Pull request by utkarsh867**
This is meant to make the auth in the application separate from the graphql API.

### CODE REVIEW

**Comment by utkarsh867** [comment]
This aims to deprecate the Login mutation in GraphQL for Phoenix. The Register mutation will still stay in GraphQL because it just makes sense. However, for login, it makes little sense to me, to be using GraphQL.

@98sean98 I think you will want to take a look at the changes and try to modify Ladybird accordingly.

**Comment by utkarsh867** [comment]
@98sean98 Major additional features to Phoenix!

It now has support to handle Admin users in the best way I could come up with.

Admin users can login and register using REST API and the token can be used to access GraphQL. Seems like a neat implementation.

**Comment by 98sean98** [comment]
I see. We can drop #44 in favour of this instead. I'll spend a bit of time to go through the code base in detail for this.

File: src/utils/resolveUser.ts [hunk 1]
```diff
@@ -1,40 +1,42 @@
...
+      const adminSession = await prisma.adminUserSession({
+        id: sessionObject.id,
+      });
+      if (adminSession === null) {
+        return null;
+      } else {
+        const userObject = await prisma
+          .adminUserSession({ id: sessionObject.id })
+          .user();
+        return { user: userObject, priviledge: "admin" };
```
Comment (98sean98) [inline src/utils/resolveUser.ts:30]: its spelled `privilege` not `priviledge` right?

File: src/server.ts [hunk 2]
```diff
@@ -4,8 +4,10 @@ env.config();
 
 import typeDefs from "./schema";
 import resolvers from "./resolvers";
-import { prisma } from "./generated/prisma-client";
+import { prisma, Admin } from "./generated/prisma-client";
 import Cookies from "universal-cookie";
+import { resolveUserUsingJWT } from "./utils/resolveUser";
+import { User } from "./generated/prisma-client";
```
Comment (98sean98) [inline src/server.ts:10]: needs refactor

File: src/server.ts [hunk 3]
```diff
@@ -16,16 +18,25 @@ import Cookies from "universal-cookie";
...
-  context: ({ req }): object => {
-    // to get token from cookies
+  context: async ({ req }): Promise<object> => {
     const cookies = new Cookies(req && req.headers.cookie);
-    const cookieToken = cookies.get("RCTC_USER");
-    // token for backwards compatibility
+    const cookieToken: string = cookies.get("RCTC_USER");
     const fallbackToken = req && req.headers.authorization;
-
+    const token: string = cookieToken ? cookieToken : fallbackToken;
```
Comment (98sean98) [inline src/server.ts:25]: change `cookieToken ? cookieToken : fallbackToken` to `cookieToken || fallbackToken`

File: src/server.ts [hunk 4]
```diff
@@ -16,16 +18,25 @@ import Cookies from "universal-cookie";
...
       prisma,
-      token: cookieToken ? cookieToken : fallbackToken,
+      token,
+      auth: {
+        user: priviledge === "user" ? user : null,
+        isAuthenticated: priviledge === "user" && user !== null,
+        isAdmin: priviledge === "admin",
+        admin: priviledge === "admin" ? user : null,
+        isAdminAuthenticated:
+          priviledge === "admin" && user !== null ? user : null,
```
Comment (98sean98) [inline src/server.ts:38]: `isAdminAuthenticated` should be a `boolean` right? which should be just `privilege === 'admin' && user !== null`. Also, I feel some of these fields a little redundant? like maybe `isAuthenticated` and `isAdminAuthenticated` can be combined, basically what we need to achieve is to check if a certain http request is authenticated or not, is it from a `user` or an `admin`, and which `user` or `admin`

File: src/mutations/events/events.ts [hunk 5]
```diff
@@ -31,11 +31,12 @@ const createEvent = async (
     image_url: "",
     organiser: {
       connect: {
-        id: user.id,
+        username: user.username,
       },
```
Comment (98sean98) [inline src/mutations/events/events.ts:35]: umm why this change here? shouldn't `id` always exist on `user`?

File: src/auth/user.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        .status(401)
+        .send(
+          "Not using valid email address. Check if the email address ends with @connect.hku.hk"
+        );
+      return;
+    }
+
+    try {
+      const user: User = await prisma.user({ email: payload.email });
+      if (user == null) {
```
Comment (98sean98) [inline src/auth/user.ts:34]: should use `===`

File: src/auth/user.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  } catch (e) {
+    console.error(e);
+    res
+      .status(500)
+      .send(
+        "There was an error at the server when making the Google verification request"
+      );
+    return;
+  }
```
Comment (98sean98) [inline src/auth/user.ts:87]: 4 `try catch` blocks put together, feels like callback hell lol. can't we separate each one to make code more debuggable in general?

File: src/utils/resolveUser.ts [hunk 8]
```diff
@@ -1,40 +1,42 @@
...
   try {
-    const userSession: UserSessions = jwt.verify(
-      ctx.token,
+    const sessionObject = jwt.verify(
+      token,
       process.env.PRISMA_SECRET
-    ) as UserSessions;
-    const sessionId = userSession.id;
-    return ctx.prisma.userSessions({ id: sessionId }).user();
+    ) as SessionObject;
```
Comment (98sean98) [inline src/utils/resolveUser.ts:18]: Please walk me through this. `generateToken()` it takes in either a `UserSession` or an `AdminUserSession` to sign a `token`. So when a `token` is verified and converted back, it can be either of them, but is casted as a `SessionObject` where the `id` is important, and used to identify if the token is from a `UserSession` or `AdminUserSession` by searching the database.

Comment (98sean98) [reply src/utils/resolveUser.ts:18]: I was thinking shouldn't it be more elegant to use just a singular `Session` datamodel where it has both a `user` and `admin` field (`@relation` with either a `User` or an `Admin`, both nullable). And then check which one is not `undefined`, so we could tell what kind of `Session` this is. I just find it funny why should 2 kinds of `Session`s exist when they serve the same purpose.

**Comment by 98sean98** [comment]
With the REST api, [react-admin](https://marmelab.com/react-admin/) does make it faster to build a dashboard for the hall admins. It might make sense to just bundle all of our applications together so that the hall has access to everything in one place.

File: src/mutations/events/events.ts (hunk 5 above)
Comment (utkarsh867) [reply src/mutations/events/events.ts:35]: Oh this was definitely an experimental change. Going to revert this.

File: src/auth/user.ts (hunk 7 above)
Comment (utkarsh867) [reply src/auth/user.ts:87]: Yeah, I felt these were excessive too. Figuring out how to reduce the nesting. Next commit.

File: src/utils/resolveUser.ts (hunk 8 above)
Comment (utkarsh867) [reply src/utils/resolveUser.ts:18]: The issue I think is that Prisma does not allow storing `User | Admin`. And I think that adding a `user` and `admin` in one table defeats the purpose of trying to unify `Session` anyway? The logic remains the same where you will check whether it is an admin or user? Am I missing something? Let me know your idea in detail.

File: src/server.ts (hunk 4 above)
Comment (utkarsh867) [reply src/server.ts:38]: Oh yes, my bad

Comment (utkarsh867) [reply src/server.ts:38]: I think having a `isAdminAuthenticated` is helpful, even though it seems a little redundant right now.

We avoid having to check if `user===null` and so on. We just check if admin is authenticated then use the admin object without worrying about whether its null, or whether user is not null etc.

File: src/utils/resolveUser.ts (hunk 1 above)
Comment (utkarsh867) [reply src/utils/resolveUser.ts:30]: oops

**Comment by utkarsh867** [comment]
@98sean98 I think what I have done here is I have tried to convince myself that `Admin` and `User` can in fact use the same datamodel. And it seems to be working!

Please review and run some checks on your end to make sure this works. I strongly believe that it does.

**Comment by utkarsh867** [comment]
@98sean98 Waiting for another review! Made the changes from the last review. It's quite a long changelog 😛 Sorry for that ahahha

**Comment by utkarsh867** [comment]
@welvin21 @98sean98 Please note that tests for bookings have been added. They check for:
1. User can modify their own bookings
2. User cannot modify other user's bookings.
3. User can delete their own bookings.
4. User cannot delete other user's bookings.

**Comment by utkarsh867** [comment]
@98sean98 And the changelog keeps growing...

**Comment by welvin21** [comment]
> @welvin21 @98sean98 Please note that tests for bookings have been added. They check for:
> 
> 1. User can modify their own bookings
> 2. User cannot modify other user's bookings.
> 3. User can delete their own bookings.
> 4. User cannot delete other user's bookings.

will extend `events` test code based on this as reference once this is merged

File: src/auth/user.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  const token: string = req.headers.authorization;
+  const payload = await getPayloadFromGoogle(token);
+  if (payload === null) {
+    res
+      .status(401)
+      .send(
+        "Not using valid email address. Check if the email address ends with @connect.hku.hk"
+      );
+    return;
+  }
```
Comment (98sean98) [inline src/auth/user.ts:39]: this doesn't seem right, no checks for `@connect.hku.hk` were done before this error response

**Comment by utkarsh867** [comment]
There is a really important change in the server code refactoring.

Until now, we were making a separate Apollo server during testing because the `apollo-server-testing` does not allow passing headers with the request.

The solution that I think solves the problem for us in a very neat way is injecting the req object into the context, and then using the context object to initialise the server. I have not seen this way of tackling this issue on the internet, but I think it does the job in a really neat way for us. I would want to explore any downsides to this approach later.

File: src/context.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+export type AppContext = {
+  prisma: Prisma;
+  token: string;
+  auth: {
+    user: User;
+    isAuthenticated: boolean;
+  };
+};
+
+export const context = async ({ req }): Promise<AppContext> => {
```
Comment (98sean98) [inline src/context.ts:14]: maybe should rename this function as `createContext`

File: src/context.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  const cookies = new Cookies(req && req.headers && req.headers.cookie);
```
Comment (98sean98) [inline src/context.ts:15]: Is it possible for `new Cookies(req && req.headers && req.headers.cookie)` to fail with `req = null` or `req.headers = null` cases? Find it a little hard to determine if `new Cookies(null)` is a good idea

File: src/auth/user.ts (hunk 9 above)
Comment (utkarsh867) [reply src/auth/user.ts]: I should have put this in TODO.

Something new coming up here is an environment variable for declaring the organisation domain. It makes the application a little more robust, and will be helpful in the future for Google OAuth tests

Comment (98sean98) [reply src/auth/user.ts]: I see. That's great

File: src/context.ts (hunk 11 above)
Comment (utkarsh867) [reply src/context.ts:15]: It will certainly not. The tests in fact run this context object. It just falls back to the fallback token.

File: src/context.ts (hunk 10 above)
Comment (utkarsh867) [reply src/context.ts:14]: I did give this a good thought. Although before I give my opinions, I would like to know why you say that?

Comment (98sean98) [reply src/context.ts:14]: I don’t know. This function creates and returns a context. So what do you think?

File: src/auth/user.ts [hunk 12]
```diff
@@ -29,7 +29,10 @@ async function getPayloadFromGoogle(
 router.post("/login", async (req, res) => {
   const token: string = req.headers.authorization;
   const payload = await getPayloadFromGoogle(token);
-  if (payload === null) {
+  if (
+    payload === null ||
+    payload.hd !== (process.env.ORG_DOMAIN || "connect.hku.hk")
```
Comment (utkarsh867) [inline src/auth/user.ts:34]: @98sean98 I have added this here

File: src/context.ts (hunk 10 above)
Comment (98sean98) [reply src/context.ts:14]: Also there’s `AppContext` being exported. Imagine that.

Comment (utkarsh867) [reply src/context.ts:14]: I see.

I will put this differently. At first glance, it does look like a function that is returning a context. However, it is doing something different.

Think of what the context type is in the the Apollo server docs:
https://github.com/apollographql/apollo-server/blob/71361880b6ba284ee8d16d5091945a999b005703/packages/apollo-server-express/src/ApolloServer.ts#L83

It says `ContextFunction<ExpressContext, Context> | Context`. What we do here is pass a context function as our context.

So it is not returning context, it is the context. Hope this explains things enough.

File: __tests__/utils/bookings.ts [hunk 13]
```diff
@@ -4,6 +4,7 @@ import {
   Booking,
   prisma,
 } from "../../src/generated/prisma-client";
+import { bookings } from "../../src/query/bookings";
```
Comment (98sean98) [inline __tests__/utils/bookings.ts:7]: uh what is this doing here?

File: __tests__/bookings.ts [hunk 14]
```diff
@@ -139,14 +153,167 @@ describe("Booking validation", () => {
...
+          remark: $remark
+        ) {
+          id
+          remark
+        }
+      }
+    `;
+    const testUpdatedBookingInfo = {
+      id: booking.id,
+      user: "test123",
```
Comment (98sean98) [inline __tests__/bookings.ts:199]: `user` should not be needed here

File: __tests__/bookings.ts [hunk 15]
```diff
@@ -139,14 +153,167 @@ describe("Booking validation", () => {
...
+          end: $end
+          remark: $remark
+        ) {
+          id
+        }
+      }
+    `;
+    const testUpdatedBookingInfo = {
+      id: booking.id,
+      user: "test123",
```
Comment (98sean98) [inline __tests__/bookings.ts:251]: same as before

File: src/context.ts (hunk 10 above)
Comment (98sean98) [reply src/context.ts:14]: I see. I was thinking about this too.

**Comment by 98sean98** [comment]
> There is a really important change in the server code refactoring.
> 
> Until now, we were making a separate Apollo server during testing because the `apollo-server-testing` does not allow passing headers with the request.
> 
> The solution that I think solves the problem for us in a very neat way is injecting the req object into the context, and then using the context object to initialise the server. I have not seen this way of tackling this issue on the internet, but I think it does the job in a really neat way for us. I would want to explore any downsides to this approach later.

First of all, any increased security risks?

**Comment by utkarsh867** [comment]
> > There is a really important change in the server code refactoring.
> 
> > 
> 
> > Until now, we were making a separate Apollo server during testing because the `apollo-server-testing` does not allow passing headers with the request.
> 
> > 
> 
> > The solution that I think solves the problem for us in a very neat way is injecting the req object into the context, and then using the context object to initialise the server. I have not seen this way of tackling this issue on the internet, but I think it does the job in a really neat way for us. I would want to explore any downsides to this approach later.
> 
> 
> 
> First of all, any increased security risks?

Do you see any?

**Comment by utkarsh867** [comment]
> I do kinda understand the new structure now. This should make query and mutation resolvers less verbose by implementing a `permissions` layer using `graphql-shield` middleware. I think the `__tests__` need to be refactored a little; there are too many `testUserInfo` instances everywhere, when in fact they can be the same and just be exported from a file. But overall the entire request-response cycle should look a lot more streamlined and clear, without all that `isAuthenticated`, `isAdminAuthenticated`, `user`, `admin` fiasco.

I fixed the tests that I wrote.

I have left the other tests for you guys to fix

**Comment by utkarsh867** [comment]
> > > There is a really important change in the server code refactoring.
> > 
> > > 
> > 
> > > Until now, we were making a separate Apollo server during testing because the `apollo-server-testing` does not allow passing headers with the request.
> > 
> > > 
> > 
> > > The solution that I think solves the problem for us in a very neat way is injecting the req object into the context, and then using the context object to initialise the server. I have not seen this way of tackling this issue on the internet, but I think it does the job in a really neat way for us. I would want to explore any downsides to this approach later.
> > 
> > 
> > 
> > First of all, any increased security risks?
> 
> Do you see any?

Also, I am not sure why you would be concerned by it?

All I have done is decoupled the server and context. It is quite a common process called dependency injection. There is fundamentally no difference in the previous code and this code, just a different approach.

File: src/mutations/bookings.ts [hunk 16]
```diff
@@ -1,13 +1,14 @@
...
-  const user: User | null = await resolveUserUsingJWT(ctx);
-  assert.notStrictEqual(user, null, "No user login");
+import { AppContext } from "../context";
 
+const createBooking = async (
+  parent,
+  data,
+  ctx: AppContext
+): Promise<Booking> => {
+  const user: User = ctx.auth.user;
```
[the remark this answers is not in the document]
Comment (utkarsh867) [reply src/mutations/bookings.ts:11]: I think you answered your question there. We want to be very specific. If we have code that makes sure that user is defined, why have a vague type definition?

**Comment by 98sean98** [comment]
> > > > There is a really important change in the server code refactoring.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > Until now, we were making a separate Apollo server during testing because the `apollo-server-testing` does not allow passing headers with the request.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > The solution that I think solves the problem for us in a very neat way is injecting the req object into the context, and then using the context object to initialise the server. I have not seen this way of tackling this issue on the internet, but I think it does the job in a really neat way for us. I would want to explore any downsides to this approach later.
> > > 
> > > 
> > > First of all, any increased security risks?
> > 
> > 
> > Do you see any?
> 
> Also, I am not sure why you would be concerned by it?
> 
> All I have done is decoupled the server and context. It is quite a common process called dependency injection. There is fundamentally no difference in the previous code and this code, just a different approach.

I see, this should be alright I suppose. Just throwing security out there to make sure we're not doing anything dumb

**Comment by utkarsh867** [comment]
@98sean98 Let's do another round!

**Comment by 98sean98** [comment]
Gonna have to check if the full stack still works.

**Comment by utkarsh867** [comment]
Sure! Let's get this going. I have another branch lined up for you.

**Comment by utkarsh867** [comment]
> Gonna have to check if the full stack still works.

Also, I was wondering why this would be a reason for concern? I have made absolutely no changes to the infrastructure side of things.

If some issues still come up, we should have a big reason to worry.
